File: backend/core/src/common/access/project-access.service.ts
import { ForbiddenException, Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { In, Repository } from 'typeorm';
import { Project, ProjectMember } from '../../entities';

type ProjectRole = 'owner' | 'editor' | string;

@Injectable()
export class ProjectAccessService {
  constructor(
    @InjectRepository(Project)
    private readonly projectsRepo: Repository<Project>,
    @InjectRepository(ProjectMember)
    private readonly membersRepo: Repository<ProjectMember>,
  ) {}

  async getAccessibleProjectIds(userId: string): Promise<string[]> {
    const [owned, memberOf] = await Promise.all([
      this.projectsRepo.find({ where: { ownerId: userId }, select: { id: true } }),
      this.membersRepo.find({ where: { userId }, select: { projectId: true } as any }),
    ]);

    const ids = new Set<string>();
    for (const p of owned) ids.add(p.id);
    for (const m of memberOf) ids.add(m.projectId);
    return Array.from(ids);
  }

  async getProjectRole(userId: string, projectId: string): Promise<ProjectRole | null> {
    const project = await this.projectsRepo.findOne({ where: { id: projectId }, select: { ownerId: true } as any });
    if (!project) return null;
    if (project.ownerId === userId) return 'owner';

    const member = await this.membersRepo.findOne({
      where: { projectId, userId },
      select: { role: true } as any,
    });
    return member?.role ?? null;
  }

  async hasProjectAccess(userId: string, projectId: string): Promise<boolean> {
    const role = await this.getProjectRole(userId, projectId);
    return Boolean(role);
  }

  async assertProjectAccess(userId: string, projectId: string): Promise<void> {
    const ok = await this.hasProjectAccess(userId, projectId);
    if (!ok) {
      throw new ForbiddenException('프로젝트 접근 권한이 없습니다.');
    }
  }

  async assertProjectOwner(userId: string, projectId: string): Promise<void> {
    const role = await this.getProjectRole(userId, projectId);
    if (role !== 'owner') {
      throw new ForbiddenException('프로젝트 소유자만 수행할 수 있습니다.');
    }
  }

  async filterAccessibleProjectIds(userId: string, requestedProjectId?: string): Promise<string[]> {
    if (!requestedProjectId) {
      return this.getAccessibleProjectIds(userId);
    }

    await this.assertProjectAccess(userId, requestedProjectId);
    return [requestedProjectId];
  }

  async isMember(userId: string, projectId: string): Promise<boolean> {
    return this.membersRepo.exist({ where: { userId, projectId } });
  }

  async isOwner(userId: string, projectId: string): Promise<boolean> {
    return this.projectsRepo.exist({ where: { id: projectId, ownerId: userId } });
  }

  async findMembers(projectId: string): Promise<ProjectMember[]> {
    return this.membersRepo.find({ where: { projectId } });
  }

  async addMember(projectId: string, userId: string, role = 'editor'): Promise<ProjectMember> {
    const existing = await this.membersRepo.findOne({ where: { projectId, userId } });
    if (existing) return existing;
    const member = this.membersRepo.create({ projectId, userId, role } as any);
    return this.membersRepo.save(member as any);
  }

  async removeMember(projectId: string, userId: string): Promise<void> {
    await this.membersRepo.delete({ projectId, userId } as any);
  }

  async removeMemberById(projectId: string, memberId: string): Promise<void> {
    const member = await this.membersRepo.findOne({ where: { id: memberId } });
    if (!member || member.projectId !== projectId) return;
    await this.membersRepo.delete({ id: memberId } as any);
  }

  async projectsByIds(projectIds: string[]): Promise<Project[]> {
    if (projectIds.length === 0) return [];
    return this.projectsRepo.find({ where: { id: In(projectIds) } });
  }
}

File: backend/core/src/main/java/store/pjcloud/cowrite/core/service/ProjectAccessService.java
package store.pjcloud.cowrite.core.service;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.UUID;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.stereotype.Service;
import store.pjcloud.cowrite.core.entity.Project;
import store.pjcloud.cowrite.core.entity.ProjectMember;
import store.pjcloud.cowrite.core.repository.ProjectMemberRepository;
import store.pjcloud.cowrite.core.repository.ProjectRepository;

@Service
public class ProjectAccessService {
    private final ProjectRepository projectRepository;
    private final ProjectMemberRepository memberRepository;

    public ProjectAccessService(ProjectRepository projectRepository, ProjectMemberRepository memberRepository) {
        this.projectRepository = projectRepository;
        this.memberRepository = memberRepository;
    }

    public List<UUID> getAccessibleProjectIds(UUID userId) {
        List<Project> owned = projectRepository.findByOwnerId(userId);
        List<ProjectMember> memberOf = memberRepository.findByUserId(userId);

        Set<UUID> ids = new HashSet<>();
        for (Project p : owned) ids.add(p.getId());
        for (ProjectMember m : memberOf) ids.add(m.getProjectId());
        return new ArrayList<>(ids);
    }

    public String getProjectRole(UUID userId, UUID projectId) {
        Project project = projectRepository.findById(projectId).orElse(null);
        if (project == null) return null;
        if (project.getOwnerId() != null && project.getOwnerId().equals(userId)) return "owner";

        return memberRepository.findByProjectIdAndUserId(projectId, userId)
            .map(ProjectMember::getRole)
            .orElse(null);
    }

    public boolean hasProjectAccess(UUID userId, UUID projectId) {
        return getProjectRole(userId, projectId) != null;
    }

    public void assertProjectAccess(UUID userId, UUID projectId) {
        if (!hasProjectAccess(userId, projectId)) {
            throw new AccessDeniedException("프로젝트 접근 권한이 없습니다.");
        }
    }

    public void assertProjectOwner(UUID userId, UUID projectId) {
        String role = getProjectRole(userId, projectId);
        if (!"owner".equals(role)) {
            throw new AccessDeniedException("프로젝트 소유자만 수행할 수 있습니다.");
        }
    }

    public List<UUID> filterAccessibleProjectIds(UUID userId, UUID requestedProjectId) {
        if (requestedProjectId == null) {
            return getAccessibleProjectIds(userId);
        }
        assertProjectAccess(userId, requestedProjectId);
        return List.of(requestedProjectId);
    }

    public boolean isMember(UUID userId, UUID projectId) {
        return memberRepository.existsByUserIdAndProjectId(userId, projectId);
    }

    public boolean isOwner(UUID userId, UUID projectId) {
        return projectRepository.existsByIdAndOwnerId(projectId, userId);
    }

    public List<ProjectMember> findMembers(UUID projectId) {
        return memberRepository.findByProjectId(projectId);
    }

    public ProjectMember addMember(UUID projectId, UUID userId, String role) {
        return memberRepository.findByProjectIdAndUserId(projectId, userId)
            .orElseGet(() -> {
                ProjectMember member = new ProjectMember();
                member.setProjectId(projectId);
                member.setUserId(userId);
                member.setRole(role == null ? "editor" : role);
                return memberRepository.save(member);
            });
    }

    public void removeMember(UUID projectId, UUID userId) {
        memberRepository.deleteByProjectIdAndUserId(projectId, userId);
    }

    public void removeMemberById(UUID projectId, UUID memberId) {
        ProjectMember member = memberRepository.findById(memberId).orElse(null);
        if (member == null || !projectId.equals(member.getProjectId())) return;
        memberRepository.deleteById(memberId);
    }

    public List<Project> projectsByIds(List<UUID> projectIds) {
        if (projectIds == null || projectIds.isEmpty()) return List.of();
        return projectRepository.findAllById(projectIds);
    }
}
